protocol/__init__.py

```diff
@@ -0,0 +1,136 @@
+log = logging.getLogger()
+
+class Protocol():
+    """
+    Class to manage all interactions with the deployed contracts
+    """
+    def __init__(self):
+        self.datatrust = None
+
+    def init_protocol(self, rpc_path, datatrust_contract, datatrust_host, voting_contract, datatrust_key, datatrust_wallet):
+        """
+        :param rpc_path: URL to the RPC provider for the network
+        :param datatrust_contract: Deployed address of the datatrust contract
+        :param datatrust_host: URL of this application. Used to submit self as datatrust
+        :param voting_contract: Deployed address of voting contract
+        :param datatrust_key: Private key for wallet
+        :param datatrust_wallet: Wallet address
+        """
+        self.w3 = self.initialize_web3(rpc_path)
+        self.datatrust_contract = datatrust_contract
+        self.datatrust_host = datatrust_host
+        self.voting_contract = voting_contract
+        self.datatrust_key = datatrust_key
+        self.datatrust_wallet = datatrust_wallet
+
+    def initialize_web3(self, rpc_path):
+        """
+        Setup the web3 provider
+        """
+        w3 = Web3(Web3.HTTPProvider(rpc_path))
+        log.info('w3 provider set')
+        return w3
+
+    def initialize_datatrust(self):
+        """
+        Confirm or create role as datatrust backend in protocol
+        """
+        log.info('Getting current datatrust address from network')
+        self.datatrust = Datatrust(self.datatrust_wallet)
+        self.datatrust.at(self.w3, self.datatrust_contract)
+
+        backend = call(self.datatrust.get_backend_address())
+        if backend == self.datatrust_wallet:
+            log.info('This server is the datatrust host. Resolving registration')
+            resolve = send(
+                self.w3, 
+                self.datatrust_key, 
+                self.datatrust.resolve_registration(
+                    self.w3.sha3(text=self.datatrust_host), {'from': self.datatrust_wallet}
+                    )
+                )
+            log.info(f'Resolved, transaction id: {resolve}')
+        else:
+            # backend not set, or is set to a different host
+            datatrust_url = call(self.datatrust.get_backend_url())
+            if datatrust_url == self.datatrust_host:
+                log.info('Server has been registered as datatrust, but not voted in')
+            else:
+                log.info('No backend or different host set. Resolving prior registrations and Submitting this one for voting')
+                datatrust_hash = self.w3.sha3(text=self.datatrust_host)
+                #TODO: Get the hash of any pending candidates and resolve before registration
+                resolve = send(
+                    self.w3, 
+                    self.datatrust_key, 
+                    self.datatrust.resolve_registration(
+                        self.w3.sha3(text=self.datatrust_host), {'from': self.datatrust_wallet}
```

Comment:
same. likely do not need to include `from` here. in the future you likely _will_ need to include gas_price however. 

Skynet will always take `2 gwei` so its not too pressing but perhaps make a ticket for the future along the lines of _dynamically figure current gas_price_ or something... (will matter on rinkeby/mainnet)

Reply:
Gas price estimate logged as #18